nat-lab/tests/test_direct_connection.py

```diff
@@ -76,90 +92,223 @@
         ConnectionTag.DOCKER_OPEN_INTERNET_CLIENT_1,
         DOCKER_OPEN_INTERNET_CLIENT_1_IP,
     ),
+    (
+        UPNP_PROVIDER,
+        ConnectionTag.DOCKER_UPNP_CLIENT_1,
+        ConnectionTag.DOCKER_UPNP_CLIENT_2,
+        DOCKER_UPNP_CLIENT_2_IP,
+    ),
```

Comment:
We should also add UPNP provider where it is available only on one side, e.g. UPnP<->CONE

Reply:
other than that looks good -> +1.0

Reply:
That makes it +2 with Tomasz. Could you approve the PR? :)